# Assignment 6

In [7]:
# Script name: Assignment 6
# Purpose of script: Report the partial effects in a linear regression model
# Author: Minh-Thy Tyler
# Date Created: 
# 10/05/2022
# Notes: N/A
# ---------------------------

# Packages
import wooldridge as woo
import numpy as np 
import pandas as pd
import statsmodels.formula.api as smf
import os
# Set up working directory
directory_path = '/Users/minh-thytyler/Desktop/2022-2023/ECON 320/HW' # use / instead of \ or use \ but insert r before (for instance: r'C:\your-directory-path')
os.chdir(directory_path)

## Part 1

In [8]:
#Perform a data generating process of y = B0 + B1 * x1 + B2*x2 + u
birthday = '08182001'
rng = np.random.RandomState(int(birthday))

beta0=9
beta1=8
beta2=2

#scale stands for std
u = rng.normal(loc=0,scale=1**.5,size=1000000)
x1 = rng.normal(loc=10,scale=8**.5,size=1000000)
x2 = rng.normal(loc=1, scale = 8**.5, size=1000000)
y = beta0 + beta1 * x1 + beta2 * x2 + u


#data variable is in a data frame

popdata_reg = pd.DataFrame({'y' : y, 'x1' : x1, 'x2': x2, 'u':u })
print(popdata_reg.head())
print(popdata_reg.describe())

            y         x1        x2         u
0   60.108906   7.251365 -3.920379  0.938747
1   94.948370  10.071606  3.001042 -0.626558
2   79.309141   9.278553 -1.876825 -0.165628
3   76.982313   8.445753  0.063148  0.289993
4  107.448499  12.783474 -2.211435  0.603579
                    y              x1              x2               u
count  1000000.000000  1000000.000000  1000000.000000  1000000.000000
mean        90.991411        9.998426        1.002822       -0.001645
std         23.365118        2.831831        2.832203        1.001314
min        -25.564294       -3.936529      -12.446616       -4.829518
25%         75.228243        8.089034       -0.908924       -0.677874
50%         90.967606        9.995395        1.002772       -0.001898
75%        106.730188       11.907383        2.913155        0.674147
max        206.568811       24.069874       14.150682        5.224590


In [9]:
# Get a sample of 1000 observations
reg_df = pd.DataFrame(popdata_reg)
sample_df = reg_df.sample(n = 1000, random_state = int(birthday))

### Getting the estimators

In [10]:
#Estimating a linear model
mod = smf.ols(formula='y ~ x1 + x2', data=sample_df)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.998
Model:                            OLS   Adj. R-squared:                  0.998
Method:                 Least Squares   F-statistic:                 2.910e+05
Date:                Tue, 18 Oct 2022   Prob (F-statistic):               0.00
Time:                        11:49:17   Log-Likelihood:                -1395.0
No. Observations:                1000   AIC:                             2796.
Df Residuals:                     997   BIC:                             2811.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      9.0726      0.113     80.468      0.000       8.851       9.294
x1             8.0001      0.011    742.367      0.000       7.979       8.021
x2             1.9921      0.011    186.298      0.000       1.971       2.013
==============================================================================
Omnibus:                        1.062   Durbin-Watson:                   1.993
Prob(Omnibus):                  0.588   Jarque-Bera (JB):                1.096
Skew:                          -0.020   Prob(JB):                        0.578
Kurtosis:                       2.843   Cond. No.                         38.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Are the estimates close or different?**

**Answer:**


The estimates are close to the population value. For my sample of 1000, I got an intercept of 9.07 (actual is 9), the beta 1 estimator is 8.0001 (actual is 8), and the beta 2 estimator is 1.9921 (actual is 2). The estimates is close to the population because we took a relatively large sample of n=1000, meaning that it will have a closer sample estimate of the actual. As we increase the sample size, the closer the estimates are to the actual in the population. 

Multiple regression is a braoder class of regressions than simple linear regression because it encompasses linear and nonlinear regressions with mulitple explanatory variables. Because of this, multiple linear regression captures multiple independent variables that influence the dependent variable and will more accurately capture the dependent variable y. 



### Partialling  out $\beta_1$ <a class = anchor id = anchor4></a>

In [22]:
#Partialling out beta1 and expressing x1 as a function of x2
mod12 = smf.ols(formula='x1 ~ x2', data=sample_df)
res12 = mod12.fit()

# Create the variable in df
x1x2_beta1 = res12.params[1]
x1x2_beta0 = res12.params[0]

sample_df['x1_hat'] = res12.predict()
#sample_df['x1_hat'] = x1x2_beta0 + x1x2_beta1*sample_df['x2']
sample_df['x1_resid'] = sample_df['x1'] - sample_df['x1_hat']
sample_df.head()

y         x1        x2         u     x1_hat  x1_resid  \
953204  157.135043  17.173613  4.538618  1.668905   9.957148  7.216464   
667466   82.145132  10.425222 -5.154320  0.051997  10.097359  0.327863   
533096   91.544631  10.303541 -0.474004  1.064308  10.029657  0.273885   
243481   86.487759   8.763398  4.398199 -1.415825   9.959180 -1.195781   
848331   90.371576  11.343264 -4.134138 -1.106260  10.082601  1.260663   

            x2_hat   x2_resid  
953204  148.651829   8.483214  
667466   93.542253 -11.397121  
533096   93.110417  -1.565786  
243481   81.352927   5.134832  
848331  101.004751 -10.633175

In [29]:
mody1 = smf.ols(formula = 'y ~ x1_resid', data = sample_df)
resy1 = mody1.fit()

resy1.params.round(2)

Intercept    90.92
x1_resid      8.00
dtype: float64

**x1_resid: 8.00** the estimate is close to the beta 1 value

### Partialling  out $\beta_2$ <a class = anchor id = anchor6></a>

In [32]:
#CHANGE 'x2 ~ x1 to y ~ x1_resid'

mod21 = smf.ols(formula='x2 ~ x1', data=sample_df)
res21 = mod21.fit()

# Create the variable in df
x2x1_beta1 = res21.params[1]
x2x1_beta0 = res21.params[0]

sample_df['x2_hat'] = res21.fittedvalues
sample_df['x2_resid'] = res21.resid
#sample_df['x2_resid'] = sample_df['x2'] - sample_df['x2_hat']
sample_df.head()

y         x1        x2         u     x1_hat  x1_resid  \
953204  157.135043  17.173613  4.538618  1.668905   9.957148  7.216464   
667466   82.145132  10.425222 -5.154320  0.051997  10.097359  0.327863   
533096   91.544631  10.303541 -0.474004  1.064308  10.029657  0.273885   
243481   86.487759   8.763398  4.398199 -1.415825   9.959180 -1.195781   
848331   90.371576  11.343264 -4.134138 -1.106260  10.082601  1.260663   

          x2_hat  x2_resid  
953204  0.781156  3.757463  
667466  0.880306 -6.034626  
533096  0.882094 -1.356098  
243481  0.904722  3.493477  
848331  0.866818 -5.000955

In [33]:
mody2 = smf.ols(formula = 'y ~ x2_resid', data = sample_df)
resy2 = mody2.fit()

resy2.params.round(2)

Intercept    90.92
x2_resid      1.99
dtype: float64

**x2_resid: 1.99** The estimate is close to the b2 value